### PR TITLE
Program support

### LINKED ISSUE #34: Program support
- Various programs have different data and different UI.
- Build support in the app to be able to use it across programs

### PR DESCRIPTION
**Pull request by sudheshna-donthineni**
This code contains logic that displays the component `IdEntry` based on the program using the auth layer. 
- The URL will now contain `program` and `authType`.  
- Based on this program name, `getProgramData` API is called to retrieve all details regarding the program. 
- Basic validation login is based on the validation type (`basicValidationMapping`)
- For advanced validation, the database and the column name is also a variable now, based on the program.

The backend PR lies here https://github.com/avantifellows/auth-layer-backend/pull/4

### CODE REVIEW

File: src/components/IdEntry.vue
Comment (pritamps) [inline src/components/IdEntry.vue]: Need a better name here. `displayText ` tells us nothing about what this is for.  Also (it has to be changed anyway, but notion and code should be consistent), according to the notion JSON, shouldn't this be just `display`?

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: Wouldn't it be better to use a more functionality-specific name that is independent of what we are using on the backend?
```suggestion
import ProgramAPIService from "@/services/API/getProgramData.js";
```

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: the convention is that it should be above the param, just as we write docstrings above functions

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: same comment on docstring as above.

**Comment by deepansh96** [comment]
A few points --
- I saw that there was no issue created and linked to this PR. We've written all the main, important steps in our development lifecycle doc [here](https://www.notion.so/avantifellows/Development-Lifecycle-9acaf86e4fce4c3081b089e1b6482349). Please make an issue and link here! It helps a lot in the long run

- Also, for this feature, PRs had to be made in two repos, `auth-layer` and `auth-layer-backend`. The best practice in these cases is to keep the names of the PRs identical. Right now, this PR is called `Different programs support` and the one in the `auth-layer-backend` repo is called `Program support`. Ideally both names should be the same. Again, this helps a lot in the long term. For us (when we forget about the code) and for others who will work on the project.

- Also, when two PRs are interlinked and depend on each other, one needs to mention that in the description.  (more about it in the doc linked above)

- These last two points were not written in the doc so I've added them. I'll keep adding things from my side regarding the process. Please make sure you go through the doc again

File: src/components/IdEntry.vue
Comment (deepansh96) [inline src/components/IdEntry.vue]: No need to write `this.` in the template part of a vue file. The template already has access to the component instance (which is referred to by `this`). 

Screenshot 2021-12-16 at 12 56 48 PM

In the docs above it tells us that `this` is required to access the component instance **only in lifecycle hooks**.

Also see this https://v3.vuejs.org/guide/template-syntax.html#text

File: src/components/IdEntry.vue
Comment (deepansh96) [inline src/components/IdEntry.vue]: Why name the component `IdEntry.vue`? I feel like just `Entry.vue` should be descriptive enough!

File: src/router/index.js
Comment (dalmia) [inline src/router/index.js]: this seems like something that should be a part of purposeParams. with purposeParams being like a dictionary.

File: src/services/API/checkUser.js
Comment (dalmia) [inline src/services/API/checkUser.js]: this is coupling backend and the frontend, right? can we decouple them please so that the only connection between them is the API endpoint?

File: src/services/validation.js
Comment (dalmia) [inline src/services/validation.js]: same point on decoupling backend and frontend as above

File: src/services/basicValidationMapping.js
Comment (dalmia) [inline src/services/basicValidationMapping.js]: I think this was mentioned in a previous PR too, but there is a slight correction in how this is done!
```suggestion
/** 
  * Mapping between basicValidationType and the basic validation login
  * @param {Object} event - event triggered when a character is typed
  */
```

File: src/services/basicValidationMapping.js
Comment (dalmia) [inline src/services/basicValidationMapping.js]: inconsistency in the spacing between `function` and `(event)` across the two functions.
I think this was mentioned in a previous PR too that it would be very helpful during reviews if all these basic issues are taken care of on the local system itself either through pre-commits or through a linter. Have you set them up?

File: src/services/basicValidationMapping.js
Comment (dalmia) [inline src/services/basicValidationMapping.js]: since this function will be imported in a different file for use, just having `type` would make it very hard to understand the type of what, no?

```suggestion
export const validationTypeToFunctionMap = {
```

File: src/services/API/sqs.js
Comment (dalmia) [inline src/services/API/sqs.js]: as mentioned in a previous comment, programName feels like it should be a part of purposeParams

File: src/services/API/getProgramData.js
Comment (dalmia) [inline src/services/API/getProgramData.js]:
```suggestion
```

File: src/services/API/getProgramData.js
Comment (dalmia) [inline src/services/API/getProgramData.js]:
```suggestion
```

File: src/services/API/getProgramData.js
Comment (dalmia) [inline src/services/API/getProgramData.js]: the description seems to be wrong

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: why `var`?

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: why not?
```suggestion
    this.textObject = {
        placeholderText: textObject["placeholder"],
        ...
    }
```

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: same suggestion as above.

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: comment for what this means? it's not clear to me what is the purpose of this.

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: suggesting a different name for the variable + that of inner params to avoid redundancy.

```suggestion
      inputParams: {
        basicValidationType: "",
        maxLength: 0,
        mode: null,
        type: null,
      },
```

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]:
```suggestion
      dataSourceParams: {
```

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: can you give some examples?

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue:61]: same suggestion as before of including it under purposeParams

File: src/components/IdEntry.vue
Comment (dalmia) [inline src/components/IdEntry.vue]: can you give some examples? what is the difference between input type and input mode?

File: src/components/IdEntry.vue
Comment (dalmia) [inline src/components/IdEntry.vue]: please be a bit more descriptive with the comments so that it is immediately clear what something is doing. 

```suggestion
    <!-- button to add another input -->
```

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]:
```suggestion
      textParams: {
```

File: src/pages/Home.vue
Comment (dalmia) [inline src/pages/Home.vue]: understood now after reading everything end-to-end what it means. a better name could be `maxInputsAllowed` or `maxEntriesAllowed`

File: src/components/IdEntry.vue
Comment (dalmia) [inline src/components/IdEntry.vue]: i thought we had decided to wrap such conditions under methods/computed properties with meaningful names, right?

File: src/components/IdEntry.vue
Comment (dalmia) [inline src/components/IdEntry.vue]: can we use a better name?
like `Input.vue` or `InputForm.vue`?

File: src/components/IdEntry.vue
Comment (dalmia) [inline src/components/IdEntry.vue]:
```suggestion
     * - no ID has been entered
```

File: src/components/IdEntry.vue
Comment (dalmia) [inline src/components/IdEntry.vue]:
```suggestion
     * - ID hasn't been completely entered
```

File: src/components/IdEntry.vue
Comment (dalmia) [inline src/components/IdEntry.vue]: need to handle the case when the validation type is not present in the mapping

File: src/components/IdEntry.vue
Comment (dalmia) [inline src/components/IdEntry.vue]: suggestions as before on renaming all of these and restructuring some

**Review by dalmia — CHANGES_REQUESTED** [review]

File: src/services/validation.js
Comment (deepansh96) [inline src/services/validation.js]: Some of these lines are not linted. Do you have a linter set up on your VSCode for plain js files?
Even if that's not set up, you should set up pre-commit for this repo as well, locally AND as a github actions.

Set it up once and forget about linting issues forever!

File: src/services/validation.js
Comment (deepansh96) [inline src/services/validation.js]: We'll have to think this through thoda. I think we were also supposed to support different **_types_** of datastores right?

In the JSON that you've written in the doc, I also see a `type` option. (attaching a screenshot)
Screenshot 2021-12-16 at 7 41 02 PM

But I don't see any code which is checking the type and then redirecting to different methods? Althought right now we only have one type but it's not being checked anywhere. 

Ideally this file, `validation.js` should deal with all kinds of validations from all kinds of data sources. The method called `validateID` should take in the `userID`, `validateCount` and **_all the datastore details_**. 

When we're just passing `collectionName` and `columnName`, we're coupling this validation method to a firebase type datastore. I'm guessing you would've thought of that already. We'll deal with other kinds of datastore when they come but we should start building in the support! Like, in our case currently, pass the whole datastore object, check it's type, and call the `firebaseAPI.checkUserExists()` method if the type is firebase.

We'll discuss more on the call!

File: src/services/API/checkUser.js
Comment (deepansh96) [reply src/services/API/checkUser.js]: Responding to Aman's comment on this line - 
I'm not sure we have any other option? `checkUserExists` is a function deployed on cloud which needs these two pieces of information to work. What would you suggest we do?

File: src/pages/Home.vue
Comment (deepansh96) [reply src/pages/Home.vue]: Responding to Aman's comment - 
Yes agreed.
Two things to notice here
- A file is called `getProgramData`. There's a "get" in front of the name. That usually signifies a getter or a method. A file should be named by what it contains. This file contains a method that retrieves program data using an API call. So maybe we should think about renaming this file. Just program data is also fine! Because right now it contains code to GET the program data. Maybe we might decide to PUSH/UPDATE/DELETE some program data in the future. Then it won't make sense to create separate files for that code right? All of that code can come under a file called `ProgramData.js`.

- And as Aman suggested, the name should be changed. Again, the fact that it's a **_firebase_** API is something that can be changed in the future. Also, there's no need for this component (Home.vue) to know what kind of cloud provider or backend are we talking to. It just needs to know what a particular method will do.

--------

- Just sharing below what we did in Plio and why we follow that pattern everywhere. It made our lives so much easier/
- In Plio, we would have lots of different types of actions being taken on a Question object in the backend. So we made a file called `Question.js`. In that file we wrote something like
```js
export default {
    getQuestion() { ... },
    deleteQuestion() { ... },
    updateQuestion() { ... }
}
```
- What this code above is doing is exporting an _**object**_ that _**contains**_ these three methods. It's not exporting the methods directly. There's a difference. So while importing this in the parent file, we would write something like `import QuestionAPIService from "../../Question.js"`
- This `QuestionAPIService` is an _**object**_ whose keys are those methods you can see in the code above. And that is why to access a method, we would have to write `QuestionAPIService.updateQuestion(...)`

File: src/pages/Home.vue
Comment (deepansh96) [reply src/pages/Home.vue:61]: @sudheshna-donthineni I think we're gonna ditch all the other params right after this program thing is defined?

We talked about this on the call no? A little confused now because then how would we know the destination information? Oh maybe through the session thing pritam is working on?

File: src/pages/Home.vue
Comment (deepansh96) [inline src/pages/Home.vue]: What is this `default`? In the JSON doc on notion I could see no `default`?
image

File: src/pages/Home.vue
Comment (deepansh96) [reply src/pages/Home.vue]: Responding to Aman's comment - 
I don't think there's need for that as well. 

@sudheshna-donthineni  So what you've done here is you got a programData object from the API call to firebase. After getting it, you're extracting things one by one and putting them into a local variable called textObject, and then passing it to the IdEntry component.

Why not just pass the whole object directly to the IdEntry component without all this parsing? 
If there's some null handling to be done, some edge cases to manage, you can do that here but this is just picking up from one place and putting it into another no?

Same with all the other objects below. All you would need to do is change some variables names in your IdEntry.vue

File: src/router/index.js
Comment (deepansh96) [reply src/router/index.js]: Responding to aman - we can't send dictionaries as query params no?

File: src/services/API/endpoints.js
Comment (deepansh96) [inline src/services/API/endpoints.js]: Maybe not valid right now but will be soon, we have to think this over as well.

If other people can also use the auth layer, they might have a different place where they're storing their programData (or groupData) so we need to provide the support for hooking up their data store to get the program data file. @pritamps

File: src/services/basicValidationMapping.js
Comment (deepansh96) [inline src/services/basicValidationMapping.js]: Just a little refactoring can be done here!
Here, for every key, you're directly linking the function to that key.

A better way to write this would be to define two functions separately above this `export const ...` line.
These functions can have names like `isNumeric(input)` or `isAlphaNumeric(input)`. And then when those functions are defined and have a proper name, you can link those functions to they keys `numeric` and `alphanumeric` in the object that you're exporting.

Oh btw you will also be able to use `isNumeric` inside `isAlphaNumeric` function directly without repeating code :P

**Review by deepansh96 — CHANGES_REQUESTED** [review]

File: .github/workflows/pre-commit.yml
Comment (pritamps) [inline .github/workflows/pre-commit.yml:1]: This should have been a separate PR @sudheshna-donthineni . Would have quickly gotten merged by itself.

File: src/components/Entry.vue
Comment (pritamps) [inline src/components/Entry.vue]: Is there error checking that these params are present in the JSON?

File: src/components/Entry.vue
Comment (pritamps) [inline src/components/Entry.vue:119]: 1. Add error checking to make sure all required fields exists
2. Add documentation to README explaining what the required fields are, and what are optional (if any)

File: src/components/Entry.vue
Comment (pritamps) [inline src/components/Entry.vue:129]: 1. Add error checking to make sure all required fields exist in the JSON
2. Add documentation to README explaining what the required fields are, and what are optional (if any)

File: src/components/Entry.vue
Comment (pritamps) [inline src/components/Entry.vue]: Why not name the method to `maxNumberOfIds` -- both comment and variable name seem to indicate this

File: src/components/Entry.vue
Comment (pritamps) [inline src/components/Entry.vue:397]: There seem to be a lot of common code between `Entry.vue` and `OTP.vue`, no?

**Review by pritamps — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy_to_s3_staging.yml
Comment (deepansh96) [inline .github/workflows/deploy_to_s3_staging.yml]: Should this really be a secret? :o

File: src/components/Entry.vue
Comment (deepansh96) [inline src/components/Entry.vue]: computed variable?

File: src/components/Entry.vue
Comment (deepansh96) [inline src/components/Entry.vue:82]: computed variable?

File: src/components/Entry.vue
Comment (deepansh96) [inline src/components/Entry.vue:101]: programName?

File: src/components/Entry.vue
Comment (deepansh96) [inline src/components/Entry.vue]: The comment seems a little off
As it's a message, the comment should be something like "the message to show when the input being entered by the user does not match ..."

File: src/components/Entry.vue
Comment (deepansh96) [inline src/components/Entry.vue]: you forgot to use the basicValidationType computer property here

File: src/components/Entry.vue
Comment (deepansh96) [inline src/components/Entry.vue]: Also can you add this to the above if/else block?

File: src/pages/Home.vue
Comment (deepansh96) [inline src/pages/Home.vue:61]: programName?

File: src/services/basicValidationMapping.js
Comment (deepansh96) [inline src/services/basicValidationMapping.js]: I guess all this will get updated when you merge the OTP support PR into this right?

**Review by deepansh96 — CHANGES_REQUESTED** [review]

File: src/components/Entry.vue
Comment (deepansh96) [reply src/components/Entry.vue:397]: Yess the ideal scenario will be to have just one component which will render things based on the JSON

File: src/services/basicValidationMapping.js
Comment (sudheshna-donthineni) [reply src/services/basicValidationMapping.js]: Yeah.

File: src/pages/Home.vue
Comment (sudheshna-donthineni) [reply src/pages/Home.vue:61]: Decided, no? This will be the last step after merging the OTP and youtube PR.

File: .github/workflows/pre-commit.yml
Comment (sudheshna-donthineni) [reply .github/workflows/pre-commit.yml:1]: Sorry!

**Review by pritamps — APPROVED** [review]
This looks good to me! But wait for @deepansh96 to approve also!

**Pull request merged**
